app/controllers/sessions_controller.rb

```diff
@@ -0,0 +1,46 @@
+class SessionsController < ApplicationController
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    if request.env['omniauth.auth']
+      create_from_social
+    else
+      create_from_local_login
+    end
+  end
+
+  private
+
+  def create_from_local_login
+    u = User.find_by(username: params[:username], provider: LOCAL_LOGIN)
+    if u && u.authenticate(params[:password])
+      session[:user_id] = u.id
+      session[:token] = auth_hash[:credentials][:token]
+      redirect_to root_path
+    else
+      redirect_to login_path
+    end
+  end
+
+  def create_from_social
+    auth_hash = request.env['omniauth.auth']
+    ap auth_hash
+    u = User.find_by uid: auth_hash[:uid], provider: auth_hash[:provider]
+    if !u
+      u = User.new(uid: auth_hash[:uid], provider: auth_hash[:provider])
+      #Generate a random password so that has_secure_password doesn't prevent a save
+      u.password = SecureRandom.uuid
+    end
+    u.email = auth_hash[:info][:email]
+    u.username  = auth_hash[:info][:name]
```

Comment:
why can't lines 37 and 38 go inside the if statement?  It might be good to consider rolling up lines 31-40 into a find or create by